File: lib/features/sanction_management/services/sanction_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cvms_desktop/features/sanction_management/models/sanction_enums.dart';

class SanctionService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  Future<void> handleConfirmedViolation({
    required String violationId,
    required String vehicleId,
    required String confirmedByUserId,
  }) async {
    final now = DateTime.now();

    // 1️ Count confirmed violations for this vehicle
    final confirmedViolations =
        await _firestore
            .collection('violations')
            .where('vehicleId', isEqualTo: vehicleId)
            .where('status', isEqualTo: 'confirmed')
            .get();

    final offenseNumber = confirmedViolations.size;

    // 2️ Decide sanction
    if (offenseNumber == 1) {
      // Written warning → NO sanction record
      return;
    }

    if (offenseNumber == 2) {
      await _createSuspension(
        violationId: violationId,
        vehicleId: vehicleId,
        adminUserId: confirmedByUserId,
        startAt: now,
      );
      return;
    }

    if (offenseNumber >= 3) {
      await _createRevocation(
        violationId: violationId,
        vehicleId: vehicleId,
        adminUserId: confirmedByUserId,
        startAt: now,
      );
    }
  }

  // ---------------- PRIVATE HELPERS ----------------

  Future<void> _createSuspension({
    required String violationId,
    required String vehicleId,
    required String adminUserId,
    required DateTime startAt,
  }) async {
    final endAt = _addWorkingDays(startAt, 30);

    final batch = _firestore.batch();

    final sanctionRef = _firestore.collection('sanctions').doc();
    final vehicleRef = _firestore.collection('vehicles').doc(vehicleId);

    batch.set(sanctionRef, {
      'vehicleId': vehicleId,
      'violationId': violationId,
      'type': SanctionType.suspension.value,
      'status': SanctionStatus.active.value,
      'offenseNumber': 2,
      'startAt': Timestamp.fromDate(startAt),
      'endAt': Timestamp.fromDate(endAt),
      'createdBy': adminUserId,
      'createdAt': FieldValue.serverTimestamp(),
    });

    batch.update(vehicleRef, {'registrationStatus': 'suspended'});

    await batch.commit();
  }

  Future<void> _createRevocation({
    required String violationId,
    required String vehicleId,
    required String adminUserId,
    required DateTime startAt,
  }) async {
    final batch = _firestore.batch();

    final sanctionRef = _firestore.collection('sanctions').doc();
    final vehicleRef = _firestore.collection('vehicles').doc(vehicleId);

    batch.set(sanctionRef, {
      'vehicleId': vehicleId,
      'violationId': violationId,
      'type': SanctionType.revocation.value,
      'status': SanctionStatus.active.value,
      'offenseNumber': 3,
      'startAt': Timestamp.fromDate(startAt),
      'endAt': null,
      'createdBy': adminUserId,
      'createdAt': FieldValue.serverTimestamp(),
    });

    batch.update(vehicleRef, {'registrationStatus': 'revoked'});

    await batch.commit();
  }

  DateTime _addWorkingDays(DateTime start, int days) {
    var date = start;
    var added = 0;

    while (added < days) {
      date = date.add(const Duration(days: 1));
      if (date.weekday <= DateTime.friday) {
        added++;
      }
    }
    return date;
  }
}

File: backend/src/services/sanction.service.js
const { admin, db } = require("../firebase/admin");
const { addWorkingDays } = require("../utils/date.utils");

async function applySanctionForViolation(violationId) {
  await db.runTransaction(async (tx) => {
    const violationRef = db.collection("violations").doc(violationId);
    const violationSnap = await tx.get(violationRef);

    if (!violationSnap.exists) {
      throw new Error("Violation not found");
    }

    const violation = violationSnap.data();
    if (violation.status === "confirmed") return;

    const vehicleId = violation.vehicleId;

    // Confirm violation
    tx.update(violationRef, {
      status: "confirmed",
      confirmedAt: admin.firestore.FieldValue.serverTimestamp(),
    });

    // Count confirmed violations
    const offensesSnap = await tx.get(
      db
        .collection("violations")
        .where("vehicleId", "==", vehicleId)
        .where("status", "==", "confirmed")
    );

    const offenseCount = offensesSnap.size;

    let sanctionType = null;
    let vehicleStatus = "active";
    let endAt = null;

    if (offenseCount === 1) {
      sanctionType = "warning";
    } else if (offenseCount === 2) {
      sanctionType = "suspension";
      vehicleStatus = "suspended";
      endAt = addWorkingDays(new Date(), 30);
    } else if (offenseCount >= 3) {
      sanctionType = "revocation";
      vehicleStatus = "revoked";
    }

    if (!sanctionType) return;

    const sanctionRef = db.collection("sanctions").doc();

    tx.set(sanctionRef, {
      vehicleId,
      violationId,
      type: sanctionType,
      status: "active",
      offenseNumber: offenseCount,
      startAt: admin.firestore.FieldValue.serverTimestamp(),
      endAt: endAt
        ? admin.firestore.Timestamp.fromDate(endAt)
        : null,
      createdAt: admin.firestore.FieldValue.serverTimestamp(),
    });

    tx.update(db.collection("vehicles").doc(vehicleId), {
      registrationStatus: vehicleStatus,
    });
  });
}

module.exports = { applySanctionForViolation };
